File: src/main/java/leetcode/three_sum/ThreeSum.java
package leetcode.three_sum;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public class ThreeSum {

    public static List<Integer[]> threeSum(Integer[] items) {
        var result = new ArrayList<Integer[]>();
        Arrays.sort(items);
        for (int i = 0; i < items.length; i++) {
            var firstElement = items[i];
            if (i > 0 && firstElement.equals(items[i - 1]))
                continue;

            int[] index = twoSum(items, i + 1, items.length - 1, -firstElement);

            if (index != null)
                result.add(new Integer[]{items[i], items[index[0]], items[index[1]]});
        }
        return result;
    }

    private static int[] twoSum(Integer[] items, int leftPointer, int rightPointer, int target) {
        while (leftPointer < rightPointer) {
            int value = items[leftPointer] + items[rightPointer];
            if (value < target)
                leftPointer++;
            else if (value > target)
                rightPointer--;
            else
                return new int[]{leftPointer, rightPointer};
        }
        return null;
    }


}

File: src/main/java/leetcode/three_sum/ThreeSum.kt
package leetcode.three_sum

fun threeSum(items: Array<Int>): List<Array<Int>> {
    val result = ArrayList<Array<Int>>()
    items.sort()

    items.forEachIndexed { index, element ->
        if (index > 0 && element == items[index - 1])
            return@forEachIndexed
        val indexes = twoSum(items, index + 1, items.size - 1, element)
        if (indexes != null)
            result.add(arrayOf(element, items[indexes[0]], items[indexes[1]]))
    }
    return result;
}

fun twoSum(items: Array<Int>, left: Int, right: Int, target: Int): Array<Int>? {
    var leftPointer = left
    var rightPointer = right
    while (leftPointer < rightPointer) {
        val value = items[left] + items[right]
        if (value < target)
            leftPointer++
        else if (value > target)
            rightPointer++
        else
            return arrayOf(leftPointer, rightPointer)
    }
    return null;
}